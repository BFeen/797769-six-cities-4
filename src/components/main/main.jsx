import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import PlaceCardList from "../place-card-list/place-card-list.jsx";
import CitiesList from "../cities-list/cities-list.jsx";
import Map from "../map/map.jsx";
import MainEmpty from "../main-empty/main-empty.jsx";
import Sorting from "../sorting/sorting.jsx";
import Header from "../header/header.jsx";
import offerPropTypes from "../../prop-types/offer-prop-types.js";
import cityPropTypes from "../../prop-types/city-prop-types.js";
import {ActionCreator} from "../../reducer/application/application.js";
import {getSortedOffers} from "../../common/utils.js";
import withActiveItem from "../../hocs/with-active-item/with-active-item.js";
import {getSortType} from "../../reducer/application/selectors.js";
import {ClassNames, ScreenType} from "../../common/const.js";


const PlaceCardListWrapped = withActiveItem(PlaceCardList);
const CitiesListWrapped = withActiveItem(CitiesList);

const Main = (props) => {
  const {
    sortType,
    city,
    offers,
    onCardTitleClick,
    handleCityChange,
    handleSortTypeChange,
    onCardMouseEnter,
    onCardMouseLeave,
    activeCard,
  } = props;

  const placesCount = offers.length;
  const isEmpty = placesCount === 0;
  const sortedOffers = getSortedOffers(offers, sortType);
  const {MapClassNames} = ClassNames;

  return (
    <div className="page page--gray page--main">

      <Header />

      <main className={`page__main page__main--index ${isEmpty ? `page__main--index-empty` : ``}`}>
        <h1 className="visually-hidden">Cities</h1>
        <div className="tabs">
          <section className="locations container">

            <CitiesListWrapped
              currentCity={city}
              onItemClick={handleCityChange}
            />

          </section>
        </div>
        <div className="cities">

          {isEmpty
            ? <MainEmpty />
            : <div className="cities__places-container container">
              <section className="cities__places places">
                <h2 className="visually-hidden">Places</h2>
                <b className="places__found">{placesCount} places to stay in {city.name}</b>
                <form className="places__sorting" action="#" method="get">
                  <span className="places__sorting-caption">Sort by</span>

                  <Sorting
                    onSortTypeChange={handleSortTypeChange}
                  />

                </form>

                <PlaceCardListWrapped
                  offers={sortedOffers}
                  screenType={ScreenType.MAIN}
                  onItemClick={onCardTitleClick}
                  onCardMouseEnter={onCardMouseEnter}
                  onCardMouseLeave={onCardMouseLeave}
                />

              </section>
              <div className="cities__right-section">
                <Map
                  city={city}
                  mapClassName={MapClassNames.CITIES}
                  offers={offers}
                  activeCard={activeCard}
                />
              </div>
            </div>
          }

        </div>
      </main>
    </div>
  );
};

Main.propTypes = {
  city: cityPropTypes,
  offers: PropTypes.arrayOf(offerPropTypes).isRequired,
  sortType: PropTypes.string.isRequired,
  onCardTitleClick: PropTypes.func.isRequired,
  handleCityChange: PropTypes.func.isRequired,
  handleSortTypeChange: PropTypes.func.isRequired,
  onCardMouseEnter: PropTypes.func.isRequired,
  onCardMouseLeave: PropTypes.func.isRequired,
  activeCard: PropTypes.oneOfType([
    offerPropTypes,
    PropTypes.object.isRequired,
  ]).isRequired,
};

const mapStateToProps = (state) => ({
  sortType: getSortType(state),
});

const mapDispatchToPtops = (dispatch) => ({
  handleCityChange(city) {
    dispatch(ActionCreator.changeCity(city));
  },

  handleSortTypeChange(sortType) {
    dispatch(ActionCreator.changeSortType(sortType));
  },
});

export {Main};
export default connect(mapStateToProps, mapDispatchToPtops)(Main);
