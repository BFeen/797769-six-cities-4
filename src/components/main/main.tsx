import * as React from "react";
import {connect} from "react-redux";
import PlaceCardList from "../place-card-list/place-card-list";
import CitiesList from "../cities-list/cities-list";
import Map from "../map/map";
import MainEmpty from "../main-empty/main-empty";
import Sorting from "../sorting/sorting";
import Header from "../header/header";
import {IOffer, ICity} from "../../common/types";
import {ActionCreator} from "../../reducer/application/application";
import {getSortedOffers} from "../../common/utils";
import withActiveItem from "../../hocs/with-active-item/with-active-item";
import withSorting from "../../hocs/with-sorting/with-sorting";
import {getSortType} from "../../reducer/application/selectors";
import {ClassNames, ScreenType} from "../../common/const";




interface Props {
  errorMessage: string;
  isAuthorized: boolean;
  city: ICity,
  offers: IOffer[];
  sortType: string;
  activeCard?: IOffer;
  onCardTitleClick: (offer: IOffer) => void;
  handleCityChange: (city: ICity) => void;
  handleSortTypeChange: (sortType: string) => void;
  onCardMouseEnter: (offer: IOffer) => void;
  onCardMouseLeave: () => void;
  onBookmarkClick: (offerId: number, status: number) => void;
}

const SortingWrapped = withSorting(Sorting);
const PlaceCardListWrapped = withActiveItem(PlaceCardList);
const CitiesListWrapped = withActiveItem(CitiesList);

const Main: React.FunctionComponent<Props> = (props: Props) => {
  const {
    errorMessage,
    isAuthorized,
    sortType,
    city,
    offers,
    onCardTitleClick,
    handleCityChange,
    handleSortTypeChange,
    onCardMouseEnter,
    onCardMouseLeave,
    activeCard,
    onBookmarkClick,
  } = props;

  const placesCount = offers.length;
  const isEmpty = placesCount === 0;
  const sortedOffers = getSortedOffers(offers, sortType);
  const {MapClassNames} = ClassNames;

  return (
    <div className="page page--gray page--main">

      <Header />

      <main className={`page__main page__main--index ${isEmpty ? `page__main--index-empty` : ``}`}>
        <h1 className="visually-hidden">Cities</h1>
        <div className="tabs">
          <section className="locations container">

            <CitiesListWrapped
              currentCity={city}
              onItemClick={handleCityChange}
            />

          </section>
        </div>
        <div className="cities">

          {isEmpty || errorMessage
            ? <MainEmpty errorMessage={errorMessage} city={city.name}/>
            : <div className="cities__places-container container">
              <section className="cities__places places">
                <h2 className="visually-hidden">Places</h2>
                <b className="places__found">{placesCount} places to stay in {city.name}</b>

                <SortingWrapped
                  currentSortType={sortType}
                  onSortTypeChange={handleSortTypeChange}
                />

                <PlaceCardListWrapped
                  isAuthorized={isAuthorized}
                  offers={sortedOffers}
                  screenType={ScreenType.MAIN}
                  onItemClick={onCardTitleClick}
                  onCardMouseEnter={onCardMouseEnter}
                  onCardMouseLeave={onCardMouseLeave}
                  onBookmarkClick={onBookmarkClick}
                />

              </section>
              <div className="cities__right-section">
                <Map
                  city={city}
                  mapClassName={MapClassNames.CITIES}
                  offers={offers}
                  activeCard={activeCard}
                />
              </div>
            </div>
          }

        </div>
      </main>
    </div>
  );
};

const mapStateToProps = (state) => ({
  sortType: getSortType(state),
});

const mapDispatchToPtops = (dispatch) => ({
  handleCityChange(city) {
    dispatch(ActionCreator.changeCity(city));
    dispatch(ActionCreator.resetSortType());
  },

  handleSortTypeChange(sortType) {
    dispatch(ActionCreator.changeSortType(sortType));
  },
});

export {Main};
export default connect(mapStateToProps, mapDispatchToPtops)(Main);
