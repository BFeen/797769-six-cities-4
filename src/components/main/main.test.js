import React from "react";
import renderer from "react-test-renderer";
import {Main} from "./main.jsx";


const offersMock = [
  {
    id: 0,
    type: `apartment`,
    city: `Amsterdam`,
    title: `Beautiful & luxurious apartment at great location`,
    picture: `img/apartment-01.jpg`,
    price: 200,
    isPremium: true,
    isFavorite: false,
    rating: 4,
    coordinates: [
      52.3909553943508,
      4.85309666406198
    ],
    details: {
      description: `A quiet cozy and picturesque that hides behind a a river...`,
      bedroomsCount: 3,
      pictures: [
        `img/apartment-01.jpg`,
        `img/room.jpg`,
      ],
      maxGuests: 4,
      insideItems: [
        `Wi-Fi`,
        `Heating`,
        `Kitchen`,
      ],
      host: {
        id: 0,
        avatar: `img/avatar-angelina.jpg`,
        name: `Angelina`,
        isPro: true,
      }
    }
  }, {
    id: 1,
    type: `house`,
    city: `Amsterdam`,
    title: `Wood and stone place`,
    picture: `img/apartment-03.jpg`,
    price: 170,
    isPremium: true,
    isFavorite: false,
    rating: 5,
    coordinates: [
      52.369553943508,
      4.85309666406198
    ],
    details: {
      description: `Lorem ipsum dolor, sit amet consectetur adipisicing elit.`,
      bedroomsCount: 3,
      pictures: [
        `img/room.jpg`,
        `img/apartment-02.jpg`,
      ],
      maxGuests: 4,
      insideItems: [
        `Wi-Fi`,
        `Heating`,
        `Kitchen`,
        `Washing machine`,
      ],
      host: {
        id: 1,
        avatar: `img/avatar-max.jpg`,
        name: `Max`,
        isPro: true,
      }
    }
  }, {
    id: 2,
    city: `Amsterdam`,
    title: `Canal view Princengracht`,
    picture: `img/room.jpg`,
    isPremium: false,
    isFavorite: true,
    price: 70,
    type: `room`,
    rating: 3,
    coordinates: [
      52.3909553943508,
      4.929309666406198
    ],
    details: {
      description: `Eligendi repellendus ut optio ad repudiandae`,
      bedroomsCount: 2,
      pictures: [
        `img/apartment-02.jpg`,
        `img/apartment-03.jpg`,
      ],
      maxGuests: 2,
      insideItems: [
        `Wi-Fi`,
        `Kitchen`,
        `Friddge`,
      ],
      host: {
        id: 2,
        avatar: `avatar.jpg`,
        name: `Alessa`,
        isPro: false,
      }
    }
  }, {
    id: 3,
    city: `Amsterdam`,
    title: `Nice, cozy, warm big bed apartment`,
    picture: `img/apartment-02.jpg`,
    isPremium: false,
    isFavorite: true,
    price: 150,
    type: `apartment`,
    rating: 4,
    coordinates: [
      52.3809553943508,
      4.939309666406198
    ],
    details: {
      description: `Quia labore atque nostrum eum repudiandae laboriosam`,
      bedroomsCount: 1,
      pictures: [
        `img/studio-01.jpg`,
        `img/apartment-01.jpg`,
      ],
      maxGuests: 2,
      insideItems: [
        `Wi-Fi`,
        `Kitchen`,
        `Friddge`,
        `Cabel TV`,
      ],
      host: {
        id: 3,
        avatar: `avatar-2.jpg`,
        name: `Kristian`,
        isPro: false,
      }
    }
  }
];

const cityMock = {
  name: `Amsterdam`,
  coordinates: [52.38333, 4.9],
};

const children = <div className="children-component" />;

describe(`Main component snapshot testing`, () => {
  it(`Render Main.`, () => {
    const tree = renderer.create(
        <Main
          offers={offersMock}
          city={cityMock}
          sortType={`popular`}
          mapClassName={`cities`}
          onCardTitleClick={() => {}}
          handleCityChange={() => {}}
          handleSortTypeChange={() => {}}
          onCardMouseEnter={() => {}}
          onCardMouseLeave={() => {}}
          activeCard={{}}
        >
          {children}
        </Main>, {
          createNodeMock: () => document.createElement(`div`)
        }
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Render Main with no offers`, () => {
    const tree = renderer.create(
        <Main
          offers={[]}
          city={cityMock}
          sortType={`popular`}
          mapClassName={`cities`}
          onCardTitleClick={() => {}}
          handleCityChange={() => {}}
          handleSortTypeChange={() => {}}
          onCardMouseEnter={() => {}}
          onCardMouseLeave={() => {}}
          activeCard={{}}
        >
          {children}
        </Main>, {
          createNodeMock: () => document.createElement(`div`)
        }
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
