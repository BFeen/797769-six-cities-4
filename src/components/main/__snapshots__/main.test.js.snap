// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component snapshot testing Render Main with no offers 1`] = `ShallowWrapper {}`;

exports[`Main component snapshot testing Render Main. 1`] = `ShallowWrapper {}`;
